### PR TITLE
112 base widget

### PR DESCRIPTION
**Pull request by MaksGMaks**
create base widget. need code review

### CODE REVIEW

File: _client/src/Ui/Component/BaseWidget.hpp [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    BaseWidget(QWidget *parent = nullptr);
+
+private:
+    // setup
+    void setupUi();
+    void setupConnections();
+
+protected:
+    // layputs
+    QVBoxLayout *_mMainLayout = new QVBoxLayout(this);
```
Comment (Sage-Cat) [inline _client/src/Ui/Component/BaseWidget.hpp]: m_mainLayout

File: _client/src/Ui/Component/BaseWidget.hpp [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    QHBoxLayout *_mButtonRow = new QHBoxLayout;
```
Comment (Sage-Cat) [inline _client/src/Ui/Component/BaseWidget.hpp]: initialization to cpp

File: _client/src/Ui/Component/BaseWidget.hpp [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    QPushButton *_mEditButton = new QPushButton("Edit", this);
+    // data table
+    QTableWidget *_mDataTable = new QTableWidget(this);
+    // additional space
+    QSpacerItem *_mAdditionalButtonSpace = new QSpacerItem(100, 50, QSizePolicy::Expanding, QSizePolicy::Fixed);
+    QSpacerItem *_mAdditionalUnderTableSpace = new QSpacerItem(100, 10, QSizePolicy::Expanding, QSizePolicy::Minimum);
+    // label
+    QLabel *_mStatus = new QLabel("Status: Ready", this);
+
+signals:
```
Comment (Sage-Cat) [inline _client/src/Ui/Component/BaseWidget.hpp]: remove redudant

File: _client/src/Ui/Component/BaseWidget.hpp [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+public slots:
```
Comment (Sage-Cat) [inline _client/src/Ui/Component/BaseWidget.hpp]: remove

File: _client/src/Ui/Component/BaseWidget.hpp [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+protected:
+    // layputs
+    QVBoxLayout *_mMainLayout = new QVBoxLayout(this);
+    QHBoxLayout *_mButtonRow = new QHBoxLayout;
+    QHBoxLayout *_mAdditionalInfo = new QHBoxLayout;
+    // buttons
+    QPushButton *_mAddButton = new QPushButton("Add", this);
+    QPushButton *_mDeleteButton = new QPushButton("Delete", this);
+    QPushButton *_mEditButton = new QPushButton("Edit", this);
```
Comment (Sage-Cat) [inline _client/src/Ui/Component/BaseWidget.hpp]: space

File: _client/src/Ui/Component/BaseWidget.hpp [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    // data table
+    QTableWidget *_mDataTable = new QTableWidget(this);
+    // additional space
+    QSpacerItem *_mAdditionalButtonSpace = new QSpacerItem(100, 50, QSizePolicy::Expanding, QSizePolicy::Fixed);
```
Comment (Sage-Cat) [inline _client/src/Ui/Component/BaseWidget.hpp]: magic number

File: _client/src/Ui/Component/BaseWidget.hpp [hunk 7]
```diff
@@ -0,0 +1,46 @@
+#ifndef BASEWIDGET_H
```
Comment (Sage-Cat) [inline _client/src/Ui/Component/BaseWidget.hpp]: pragma once

File: _client/src/Ui/Component/BaseWidget.hpp [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    QHBoxLayout *_mAdditionalInfo = new QHBoxLayout;
+    // buttons
+    QPushButton *_mAddButton = new QPushButton("Add", this);
+    QPushButton *_mDeleteButton = new QPushButton("Delete", this);
+    QPushButton *_mEditButton = new QPushButton("Edit", this);
+    // data table
+    QTableWidget *_mDataTable = new QTableWidget(this);
+    // additional space
+    QSpacerItem *_mAdditionalButtonSpace = new QSpacerItem(100, 50, QSizePolicy::Expanding, QSizePolicy::Fixed);
+    QSpacerItem *_mAdditionalUnderTableSpace = new QSpacerItem(100, 10, QSizePolicy::Expanding, QSizePolicy::Minimum);
```
Comment (Sage-Cat) [inline _client/src/Ui/Component/BaseWidget.hpp]: ADDITIONAL_UNDER_TABLE_SPACE_WIDTH

**Review by Sage-Cat — CHANGES_REQUESTED** [review]

File: _client/src/Ui/Component/BaseWidget.hpp (hunk 1 above)
Comment (MaksGMaks) [reply _client/src/Ui/Component/BaseWidget.hpp]: done

File: _client/src/Ui/Component/BaseWidget.hpp (hunk 3 above)
Comment (MaksGMaks) [reply _client/src/Ui/Component/BaseWidget.hpp]: done

File: _client/src/Ui/Component/BaseWidget.hpp (hunk 7 above)
Comment (MaksGMaks) [reply _client/src/Ui/Component/BaseWidget.hpp]: done, but not sure

File: _client/src/Ui/Component/BaseWidget.hpp (hunk 5 above)
Comment (MaksGMaks) [reply _client/src/Ui/Component/BaseWidget.hpp]: can't make. fucking visual doesn't allow it

File: _client/src/Ui/Component/BaseWidget.hpp (hunk 8 above)
Comment (MaksGMaks) [reply _client/src/Ui/Component/BaseWidget.hpp]: DONE

File: _client/src/Ui/Component/BaseWidget.hpp [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+public:
+    BaseWidget(QWidget *parent = nullptr);
+
+private:
+    // setup
+    void setupUi();
+
+protected:
+    // layputs
+    QVBoxLayout * m_MainLayout;
```
Comment (Sage-Cat) [inline _client/src/Ui/Component/BaseWidget.hpp]: correct `QVBoxLayout  *m_mainLayout;`

File: _client/src/Ui/Component/BaseWidget.hpp [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+class BaseWidget : public QWidget
+{
+    Q_OBJECT
+
+public:
+    BaseWidget(QWidget *parent = nullptr);
+
+private:
+    // setup
+    void setupUi();
```
Comment (Sage-Cat) [inline _client/src/Ui/Component/BaseWidget.hpp]: virtual

File: _client/src/Ui/Component/BaseWidget.hpp [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+const int ADDIT_SPACE_WIDTH = 100;
+const int ADDIT_SPACE_HEIGHT = 50;
+
+class BaseWidget : public QWidget
+{
+    Q_OBJECT
+
+public:
+    BaseWidget(QWidget *parent = nullptr);
```
Comment (Sage-Cat) [inline _client/src/Ui/Component/BaseWidget.hpp:17]: add virtual destructor

File: _client/src/Ui/Component/BaseWidget.hpp (hunk 10 above)
Comment (Sage-Cat) [inline _client/src/Ui/Component/BaseWidget.hpp]: protected

**Review by Sage-Cat — CHANGES_REQUESTED** [review]

File: _client/src/Ui/Component/BaseWidget.hpp [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+
+#include <QWidget>
+#include <QVBoxLayout>
+#include <QHBoxLayout>
+#include <QPushButton>
+#include <QTableWidget>
+#include <QLabel>
+#include <QSpacerItem>
+
+const int ADDIT_SPACE_WIDTH = 100;
```
Comment (Sage-Cat) [inline _client/src/Ui/Component/BaseWidget.hpp]: inside class
public

**Review by Sage-Cat — CHANGES_REQUESTED** [review]

File: _client/src/Ui/Component/BaseWidget.hpp (hunk 11 above)
Comment (MaksGMaks) [reply _client/src/Ui/Component/BaseWidget.hpp:17]: done

File: _client/src/Ui/Component/BaseWidget.hpp (hunk 10 above)
Comment (MaksGMaks) [reply _client/src/Ui/Component/BaseWidget.hpp]: done

Comment (MaksGMaks) [reply _client/src/Ui/Component/BaseWidget.hpp]: done

File: _client/src/Ui/Component/BaseWidget.hpp (hunk 9 above)
Comment (MaksGMaks) [reply _client/src/Ui/Component/BaseWidget.hpp]: done

File: _client/src/Constants.cpp [hunk 13]
```diff
@@ -0,0 +1,2 @@
+const int ADDIT_SPACE_WIDTH = 100;
```
Comment (Sage-Cat) [inline _client/src/Constants.cpp]: Add namespace with good name

File: _client/src/Constants.cpp
Comment (Sage-Cat) [inline _client/src/Constants.cpp:1]: Set better filename for spesific scope, like `WidgetStyles.hpp" or smth

**And it should be hpp!!!**

**Review by Sage-Cat — CHANGES_REQUESTED** [review]

**Review by Sage-Cat — APPROVED** [review]

**Pull request merged**